File: algorithms/56. Merge Intervals/solution.cc
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        auto cmp = [] (Interval & a, Interval & b) {return a.start < b.start;};
        sort(intervals.begin(), intervals.end(), cmp);
        vector<Interval> result;
        for ( Interval & i : intervals ) {
            if ( result.empty() || result.back().end < i.start )
                result.push_back(i);
            else
                result.back().end = max(result.back().end, i.end);
        }
        return result;
    }
};

File: algorithms/56. Merge Intervals/solution.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if not intervals:
            return []
        result = []
        intervals = sorted(intervals, key=lambda x: (x.start, x.end))
        start, end = intervals[0].start, intervals[0].end
        for i in xrange(1, len(intervals)):
            if end >= intervals[i].start:
                end = max(end, intervals[i].end)
            else:
                result.append(Interval(start, end))
                start, end = intervals[i].start, intervals[i].end
        result.append(Interval(start, end))
        return result

# 169 / 169 test cases passed.
# Status: Accepted
# Runtime: 82 ms
# beats 83.68 %
